lib/auth/helpers.go

```diff
@@ -1473,19 +1472,37 @@ type flushClt interface {
 // write events for different keys show up in the order in which the writes were performed, which
 // is not necessarily true for all backends.
 func flushCache(t *testing.T, clt flushClt) {
+	ctx := context.Background()
+
 	// the pattern of writing a resource and then waiting for it to appear
-	// works for any resource type (when using memory backend). we use namespaces
-	// here because namespaces are deprecated and therefore unlikely to interfer
-	// with tests.
+	// works for any resource type (when using memory backend).
 	name := strings.ReplaceAll(uuid.NewString(), "-", "")
-	defer clt.DeleteNamespace(name)
+	defer clt.DeleteRole(ctx, name)
+
+	role, err := types.NewRole(name, types.RoleSpecV6{})
+	if err != nil {
+		t.Fatalf("Failed to instantiate new role: %v", err)
+	}
 
-	ns, err := types.NewNamespace(name)
-	require.NoError(t, err)
+	role, err = clt.CreateRole(ctx, role)
+	if err != nil {
+		t.Fatalf("Failed to create new role: %v", err)
+	}
 
-	require.NoError(t, clt.UpsertNamespace(ns))
-	require.Eventually(t, func() bool {
-		_, err := clt.GetNamespace(name)
-		return err == nil
-	}, time.Second*20, time.Millisecond*200)
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+	for {
+		r, err := clt.GetRole(ctx, name)
+		if err == nil && r.GetRevision() == role.GetRevision() {
+			return
+		}
+
+		timer := time.NewTimer(200 * time.Millisecond)
```

Comment:
Use time.After, or maybe use a ticker defined outside the loop?

Reply:
out of curiosity, what's the benefit of using time.After vs. a timer like this?

Reply:
The main difference is that with time.NewTimer you can explicitly close the Timer, whereas with time.After you can't (because you never see the actual Timer, only the channel behind it). This used to be a bigger deal before the recent Go 1.23 timer changes - https://go.dev/doc/go1.23#timer-changes.

In production code using Timers, particularly within loops, I think I would still favor explicitly closing (although it's not as pressing as before, as I said above).

For this particular snipper I think a Ticker is what we want - it's one "timer" goroutine for the entire loop vs one timer goroutine for each iteration.

Reply:
Anything using a `clockwork.Clock` is also effectively potentially using go 1.22 timer semantics (when running in a test with a fake clock) so that's another reason to use `NewTimer` even in code using go 1.23+ - it's slightly more important when doing subtle things with timer resets however, because that can just deadlock your test as opposed to just wasting some resources.